docs/design-docs/08-package-variant.md

```diff
@@ -0,0 +1,551 @@
+
+The GVK on the other hand, has to be the same for the in-package resource and
+the in-cluster resource, because it tells us the API schema for the resource.
+Also, the namespace of the in-cluster object needs to be the same as the
+PackageVariant resource, or we could leak resources from namespaces to which
+our PackageVariant user does not have access.
+
+With that understanding, the injection process works as follows:
+
+1. The controller will examine all in-package resources, looking for those with
+   an annotation named `kpt.dev/config-injection`, with one of the following
+   values: `required` or `optional`. We will call these "injection points".
+1. For each injection point, a condition will be created in the
+   downstream PackageRevision, with ConditionType set to the dot-delimited
+   concatenation of `config.injection`, with the in-package resource kind and
+   name, and the value set to `False`. Note that since the package author
+   controls the name of the resource, kind and name are sufficient to
+   disambiguate the injection point. We will call this ConditionType the
+   "injection point ConditionType".
+1. For each required injection point, the injection point ConditionType will be
+   added to the PackageRevision `readinessGates`. Optional injection points'
+   ConditionTypes must not be added to the `readinessGates` by the
+   PackageVariant controller, but humans or other actors may do so at a later
+   date, and the package variant controller should not remove them on subsequent
+   reconciliations. Also, this relies upon `readinessGates` gating publishing
+   the package to a *deployment* repository, but not gating publishing to a
+   blueprint repository.
+1. The injection processing will proceed as follows. For each injection point:
+   - If the resource schema of the injection point is not available in the
+     cluster, then the injection point ConditionType will be set to `False`,
+     with a message indicating that the schema is missing, and processing should
+     proceed to the next injection point. Note that for `optional` injection
+     points, not having the schema may be intentional and not an error.
+   - If the resource schema of the injection point does not contain a `spec`
+     field, then the injection point ConditionType will be set to `False`, with
+     a message explaining the error, and processing should proceed to the next
+     injection point.
+   - The controller will now identify all in-cluster objects in the same
+     namespace as the PackageVariant resource, with GVK matching the injection
+     point (the in-package resource).
+   - The controller will look through the list of injection selectors in
+     order and checking if any of the in-cluster objects match the selector. If
+     so, that in-cluster object is selected, and processing of the list of
+     injection selectors stops. Note that in the initial version, the selectors
```

Comment:
So just to clarify here, if we have only one entry in the injector selector list and it only has the name specified, that will apply to all injection points specified in the package right?

And if we add information about GVK in the injector selector list entry, it will only apply if the GVK in the selector matches the GVK of the injection point?

Finally, if there is an injection point, but even after trying all the selectors in the list, we can't find a matching resource in the cluster, then the injection will not happen? And that will be true, even if we can find a unique match based only on GVK?

Reply:
I am adding some clarifying language.